cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/SparkProgramStatusMetricsProvider.java

```diff
@@ -61,53 +64,75 @@ public class SparkProgramStatusMetricsProvider implements MetricsProvider {
-        applicationResponse = doGet(applicationsURL);
-        String attemptId = extractAttemptId(applicationResponse.getResponseBodyAsString(), runIdStr);
-        if (Objects.nonNull(attemptId) && !attemptId.isEmpty()) {
-          HttpResponse stagesResponse;
-          String stagesURL = String.format("%s%s/%s/%s/stages", sparkHistoricBaseURL, sparkApplicationsEndpoint,
-                                           runIdStr, attemptId);
-          stagesResponse = doGet(stagesURL);
-          metrics = extractMetrics(stagesResponse.getResponseBodyAsString());
-          if (Objects.isNull(metrics)) {
-            logger.error("Error during metrics extraction");
-            return new ExecutionMetrics[]{};
-          } else {
-            return metrics;
+    try {
+      return Retries.supplyWithRetries(
+        () -> {
+          ExecutionMetrics[] metrics;
+          HttpResponse applicationResponse;
+          try {
+            applicationResponse = doGet(applicationsURL);
+            String attemptId = extractAttemptId(applicationResponse.getResponseBodyAsString(), runIdStr);
+            if (Objects.nonNull(attemptId) && !attemptId.isEmpty()) {
+              HttpResponse stagesResponse;
+              String stagesURL = String.format("%s%s/%s/%s/stages", sparkHistoricBaseURL, sparkApplicationsEndpoint,
+                                               runIdStr, attemptId);
+              stagesResponse = doGet(stagesURL);
+              metrics = extractMetrics(stagesResponse.getResponseBodyAsString());
+              if (Objects.isNull(metrics)) {
+                emitMetrics(runId, startTime, false);
+                logger.error("Error during metrics extraction, received null response");
+                return new ExecutionMetrics[]{};
+              } else {
+                emitMetrics(runId, startTime, true);
+                return metrics;
+              }
+            } else {
+              throw new RetryableException("Error during attemptId extraction");
+            }
+          } catch (IOException e) {
+            logger.warn("Error retrieving application response", e);
+            throw new RetryableException(e);
           }
-        } else {
-          throw new RetryableException("Error during attemptId extraction");
-        }
-      } catch (IOException e) {
-        logger.warn("Error retrieving application response", e);
-        throw new RetryableException(e);
-      }
-    }, RetryStrategies.fromConfiguration(this.cConf, Constants.Spark.SPARK_METRICS_PROVIDER_RETRY_STRATEGY_PREFIX),
-                                     t -> t instanceof RetryableException);
+        }, RetryStrategies.fromConfiguration(this.cConf, Constants.Spark.SPARK_METRICS_PROVIDER_RETRY_STRATEGY_PREFIX),
+        t -> t instanceof RetryableException);
+    } catch (Exception e) {
+      emitMetrics(runId, startTime, false);
+      logger.error("Retries failed for extracting metrics. ", e);
+      return new ExecutionMetrics[]{};
```

Comment:
We dont want to throw an exception if retries are exhausted?

Reply:
No, it is better to publish the event without metrics than throwing an exception. This is the same approach we use when we get a null response back.

Reply:
Created https://cdap.atlassian.net/browse/CDAP-19283 to improve exception handling